grpcproxy: TestGRPCStreamRequireLeader: test timed out after 15m

via https://jenkins-etcd.prod.coreos.systems/job/etcd-proxy/413/consoleFull

```
=== RUN   TestGRPCStreamRequireLeader
panic: test timed out after 15m0s
---
goroutine 36235 [chan receive, 12 minutes]:
github.com/coreos/etcd/integration.(*watchCloser).Close(0xc427325ca0, 0xc42000f000, 0xc422eb8ea0)
	/home/jenkins/workspace/etcd-proxy/gopath/src/github.com/coreos/etcd/integration/cluster_proxy.go:64 +0xa3
github.com/coreos/etcd/clientv3.(*Client).Close(0xc42071f320, 0x0, 0xc420023dc8)
	/home/jenkins/workspace/etcd-proxy/gopath/src/github.com/coreos/etcd/clientv3/client.go:90 +0x7e
github.com/coreos/etcd/integration.TestGRPCStreamRequireLeader(0xc4284aea80)
	/home/jenkins/workspace/etcd-proxy/gopath/src/github.com/coreos/etcd/integration/v3_grpc_test.go:1150 +0xe60
testing.tRunner(0xc4284aea80, 0x12b2048)
	/usr/lib/go-1.7/src/testing/testing.go:610 +0xca
created by testing.(*T).Run
	/usr/lib/go-1.7/src/testing/testing.go:646 +0x530
```

--- a/proxy/grpcproxy/watch.go
+++ b/proxy/grpcproxy/watch.go
@@ -122,23 +122,23 @@ func (wp *watchProxy) Watch(stream pb.Watch_WatchServer) (err error) {
 		defer func() { stopc <- struct{}{} }()
 		wps.sendLoop()
 	}()
-	if leaderc != nil {
-		go func() {
-			defer func() { stopc <- struct{}{} }()
-			select {
-			case <-leaderc:
-			case <-ctx.Done():
-			}
-		}()
-	}
+	// tear down watch if leader goes down or entire watch proxy is terminated
+	go func() {
+		defer func() { stopc <- struct{}{} }()
+		select {
+		case <-leaderc:
+		case <-ctx.Done():
+		case <-wp.ctx.Done():
+		}
+	}()
 
 	<-stopc
+	cancel()
+
 	// recv/send may only shutdown after function exits;
 	// goroutine notifies proxy that stream is through
 	go func() {
-		if leaderc != nil {
-			<-stopc
-		}
+		<-stopc
 		<-stopc
 		wps.close()
 		wp.wg.Done()

